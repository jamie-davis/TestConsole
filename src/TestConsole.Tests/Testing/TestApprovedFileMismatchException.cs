using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using NUnit.Framework;
using TestConsoleLib;
using TestConsoleLib.Exceptions;
using TestConsoleLib.Testing;

namespace TestConsole.Tests.Testing
{
    [TestFixture]
    public class TestApprovedFileMismatchException
    {
        [Test]
        public void ExceptionMessageIsFormatted()
        {
            //Arrange
            Exception exception = null;
            var approvedText = @"Line 1
Line 2";
            var receivedText = @"Line 1
Line 2 Diff";

            using (var approved = WriteTempFile(approvedText))
            using (var received = WriteTempFile(receivedText))
            {
                //Act
                exception = new ApprovedFileMismatchException(received.TempPath, approved.TempPath);
            }

            //Assert
            exception.Message.Verify();
        }

        [Test]
        public void ExceptionMessageHandlesEmptyReceivedFile()
        {
            //Arrange
            Exception exception = null;
            var approvedText = @"Line 1
Line 2";
            var receivedText = string.Empty;

            using (var approved = WriteTempFile(approvedText))
            using (var received = WriteTempFile(receivedText))
            {
                //Act
                exception = new ApprovedFileMismatchException(received.TempPath, approved.TempPath);
            }

            //Assert
            exception.Message.Verify();
        }

        [Test]
        public void ExceptionMessageHandlesEmptyApprovedFile()
        {
            //Arrange
            Exception exception = null;
            var approvedText  = string.Empty;
            var receivedText= @"Line 1
Line 2";

            using (var approved = WriteTempFile(approvedText))
            using (var received = WriteTempFile(receivedText))
            {
                //Act
                exception = new ApprovedFileMismatchException(received.TempPath, approved.TempPath);
            }

            //Assert
            exception.Message.Verify();
        }

        [Test]
        public void ExceptionMessageHandlesIdenticalData()
        {
            //Arrange
            Exception exception = null;
            var approvedText  = string.Empty;

            using (var approved = WriteTempFile(approvedText))
            using (var received = WriteTempFile(approvedText))
            {
                //Act
                exception = new ApprovedFileMismatchException(received.TempPath, approved.TempPath);
            }

            //Assert
            exception.Message.Verify();
        }

        [Test]
        public void MaximumDifferencesAreReported()
        {
            //Arrange
            Exception exception = null;
            var approvedText = string.Join(Environment.NewLine, Enumerable.Range(1, 100).Select(n => $"Line {n}"));
            var receivedText = string.Join(Environment.NewLine, Enumerable.Range(1, 100).Select(n => $"Line {((n & 1) != 0 ? "X" : n.ToString())}"));

            using (var approved = WriteTempFile(approvedText))
            using (var received = WriteTempFile(receivedText))
            {
                //Act
                exception = new ApprovedFileMismatchException(received.TempPath, approved.TempPath);
            }

            //Assert
            exception.Message.Verify();
        }


        #region Temporary file handling

        private TempFile WriteTempFile(string text)
        {
            return new TempFile(text);
        }

        private class TempFile : IDisposable
        {
            public string TempPath { get; }

            public TempFile(string text)
            {
                TempPath = Path.GetTempFileName();
                try
                {
                    File.WriteAllText(TempPath, text);
                }
                catch
                {
                    Dispose(); //We will not complete construction, don't leave the temp file in place
                    throw;
                }
            }

            #region IDisposable

            public void Dispose()
            {
                File.Delete(TempPath);
            }

            #endregion
        }

        #endregion
    }
}
