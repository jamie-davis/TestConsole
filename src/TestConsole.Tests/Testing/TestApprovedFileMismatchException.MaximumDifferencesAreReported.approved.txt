50 Differences found between approved and received:                                            
                                                                                               
File                   Length       Lines                                                      
-------- -------------------- -----------                                                      
Approved                  890         100                                                      
Received                  845         100                                                      
                                                                                               
File             Line Text                                                                     
-------- ------------ -------                                                                  
Approved            1 Line 1                                                                   
Received            1 Line X                                                                   
Approved            3 Line 3                                                                   
Received            3 Line X                                                                   
Approved            5 Line 5                                                                   
Received            5 Line X                                                                   
Approved            7 Line 7                                                                   
Received            7 Line X                                                                   
Approved            9 Line 9                                                                   
Received            9 Line X                                                                   
Approved           11 Line 11                                                                  
Received           11 Line X                                                                   
Approved           13 Line 13                                                                  
Received           13 Line X                                                                   
Approved           15 Line 15                                                                  
Received           15 Line X                                                                   
Approved           17 Line 17                                                                  
Received           17 Line X                                                                   
Approved           19 Line 19                                                                  
Received           19 Line X                                                                   
Approved           21 Line 21                                                                  
Received           21 Line X                                                                   
Approved           23 Line 23                                                                  
Received           23 Line X                                                                   
Approved           25 Line 25                                                                  
Received           25 Line X                                                                   
Approved           27 Line 27                                                                  
Received           27 Line X                                                                   
Approved           29 Line 29                                                                  
Received           29 Line X                                                                   
Approved           31 Line 31                                                                  
Received           31 Line X                                                                   
Approved           33 Line 33                                                                  
Received           33 Line X                                                                   
Approved           35 Line 35                                                                  
Received           35 Line X                                                                   
Approved           37 Line 37                                                                  
Received           37 Line X                                                                   
Approved           39 Line 39                                                                  
Received           39 Line X                                                                   
Approved           41 Line 41                                                                  
Received           41 Line X                                                                   
Approved           43 Line 43                                                                  
Received           43 Line X                                                                   
Approved           45 Line 45                                                                  
Received           45 Line X                                                                   
Approved           47 Line 47                                                                  
Received           47 Line X                                                                   
Approved           49 Line 49                                                                  
Received           49 Line X                                                                   
...                                                                                            
                                                                                               